experimental/gradually-typed/src/Torch/GraduallyTyped/Examples/TwoLayerNetwork.hs

```diff
@@ -0,0 +1,407 @@
+      weightDecay = 0.0,
+      amsgrad = False
+    }
+
+-- | Create a new Adam optimizer.
+mkAdam ::
+  -- | Adam options
+  AdamOptions ->
+  -- | initial model parameters
+  [ForeignPtr ATen.Tensor] ->
+  -- | Adam optimizer
+  IO (ForeignPtr ATen.Optimizer)
+mkAdam AdamOptions {..} =
+  ATen.cast7 ATen.adam learningRate beta1 beta2 epsilon weightDecay amsgrad
+
+mseLoss ::
+  forall m gradient layout device dataType shape gradient' layout' device' dataType' shape'.
+  MonadThrow m =>
+  -- | prediction tensor
+  Tensor gradient layout device dataType shape ->
+  -- | target tensor
+  Tensor gradient' layout' device' dataType' shape' ->
+  -- | output tensor
+  m
+    ( Tensor
+        (gradient <|> gradient')
+        (layout <+> layout')
+        (device <+> device')
+        (dataType <+> dataType')
+        (Seq (shape <+> shape') ('Shape '[]))
+    )
+mseLoss prediction target =
+  unsafeThrowableIO $
+    ATen.cast3
+      ATen.mse_loss_ttl
+      prediction
+      target
+      (1 :: Int) -- reduce mean
+
+computeLoss ::
+  _ =>
+  -- | device
+  SDevice device ->
+  -- | model specification
+  ModelSpec model ->
+  -- | model state dict
+  StateDict ->
+  -- | input
+  [[Float]] ->
+  -- | expected output
+  [[Float]] ->
+  -- | random generator
+  Generator generatorDevice ->
+  -- | loss
+  m (_, Generator generatorOutputDevice)
+computeLoss device modelSpec stateDict input expectedOutput =
+  runIxStateT $
+    ilift
+      ( let inputTensor = sToTensor (SGradient SWithoutGradient) (SLayout SDense) device input
+            expectedOutputTensor = sToTensor (SGradient SWithoutGradient) (SLayout SDense) device expectedOutput
+            model = flip evalStateT stateDict $ fromStateDict modelSpec mempty
+         in (,,) <$> inputTensor <*> expectedOutputTensor <*> model
+      )
+      >>>= ( \(inputTensor, expectedOutputTensor, model) ->
+               ireturn inputTensor
+                 >>>= IxStateT . forward model
+                 >>>= ilift . (`mseLoss` expectedOutputTensor)
+           )
+
+train ::
+  _ =>
+  -- | random seed for the training
+  Word64 ->
+  -- | device
+  SDevice device ->
+  -- | model specification
+  ModelSpec model ->
+  -- | model state dict
+  StateDict ->
+  -- | learning rate
+  Double ->
+  -- | stream of training examples
+  P.ListT IO ([[Float]], [[Float]], int) ->
+  IO StateDict
+train seed device modelSpec stateDict learningRate examples = do
+  g <- sMkGenerator device seed
+  optim <- mkAdam defaultAdamOptions {learningRate} (Map.elems stateDict)
+  let step g' (x, y, _iter) = do
+        let f :: ForeignPtr ATen.TensorList -> IO (ForeignPtr ATen.Tensor)
+            f tensorList = do
+              stateDict' :: StateDict <- ATen.uncast tensorList (pure . Map.fromList . zip (Map.keys stateDict))
+              UnsafeTensor t <- fst <$> computeLoss device modelSpec stateDict' x y g'
+              pure t
+        _loss :: ForeignPtr ATen.Tensor <- ATen.cast2 ATen.step optim f
```

Comment:
I haven't tested this step function very much. Please let me know if there is anything.

Reply:
Instead, `unsafeStep` has been tested.

Reply:
There's a lot going on in step. Calling the loss function, calculating the gradients, 
assuming that there's only one parameter group, running the optimizer step, returning the loss, etc.

This is fine now, but there will be times when it is not. There are features that we will need to support for training transformers that are not supported with this, including:

* gradient clipping (improves stability)
* gradient checkpointing (trades space for compute)
* gradient accumulation (increases effective batch size)

It may therefore be necessary to provide a more atomic API.

Reply:
I'm not sure I get unsafeStep. It doesn't seem to have a good signature given that the optimizer mutates the parameter tensor list.

Reply:
I also can't imagine that any of this is thread-safe. If you have two optimizers updating the same tensor list, this will certainly give you garbage